Objects/obmalloc.c

```diff
@@ -446,6 +446,7 @@ _PyMem_ArenaFree(void *Py_UNUSED(ctx), void *ptr,
 /***************************/
 
 static int
+_Py_NO_SANITIZE_THREAD
```

Comment:
I've had a better experience using suppression lists rather than marking functions as `_Py_NO_SANITIZE_THREAD`:

1) Add a file `supressions.txt` somwhere (maybe `Tools/tsan/supressions.txt`?)
2) Set the environment variable `TSAN_OPTIONS="suppressions=<path_to_supressions.txt>"`

See https://github.com/google/sanitizers/wiki/ThreadSanitizerSuppressions